packages/babel-traverse/scripts/generators/typescript.js

```diff
@@ -0,0 +1,296 @@
+    key: string): void;
+  setKey(key: string): void;
+  requeue(pathToQueue?: NodePath): void;
+
+  toComputedKey(): object;
+  ensureBlock(): t.Node;
+  arrowFunctionToShadowed(): void;
+  unwrapFunctionEnvironment(): void;
+  arrowFunctionToExpression(opts?: {
+    allowInsertArrow?: boolean,
+    specCompliant?: boolean
+  }): void;
+
+  evaluateTruthy(): boolean;
+  evaluate(): { confident: boolean; value: any };
+
+  getOpposite(): NodePath;
+  getCompletionRecords(): NodePath[];
+  getSibling(key: string): NodePath;
+  getPrevSibling(): NodePath;
+  getNextSibling(): NodePath;
+  getAllPrevSiblings(): NodePath[];
+  getAllNextSiblings(): NodePath[];
+  get<K extends keyof T>(key: K, context?: boolean | TraversalContext):
+      T[K] extends Array<t.Node | null | undefined> ? Array<NodePath<T[K][number]>> :
+      T[K] extends t.Node | null | undefined ? NodePath<T[K]> :
+      never;
+  get(key: string, context?: boolean | TraversalContext): NodePath | NodePath[];
+  getBindingIdentifiers(duplicates?: boolean): t.Node[];
+  getOuterBindingIdentifiers(duplicates?: boolean): t.Node[];
+  getBindingIdentifierPaths(duplicates?: boolean,
+    outerOnly?: boolean): NodePath[];
+  getOuterBindingIdentifierPaths(duplicates?: boolean): NodePath[];
+
+  matchesPattern(pattern: string, allowPartial?: boolean): boolean;
+  has(key: string): boolean;
+  isStatic(): boolean;
+  is(key: string): boolean;
+  isnt(key: string): boolean;
+  equals(key: string, value: any): boolean;
+  isNodeType(type: string): boolean;
+  canHaveVariableDeclarationOrExpression(): boolean;
+  canSwapBetweenExpressionAndStatement(replacement: t.Node): boolean;
+  isCompletionRecord(allowInsideFunction?: boolean): boolean;
+  isStatementOrBlock(): boolean;
+  referencesImport(moduleSource: string, importName: string): boolean;
+  getSource(): string;
+  willIMaybeExecuteBefore(path: NodePath): boolean;
+  // TODO: what type is resolved?
+  resolve(dangerous?: boolean, resolved?: object): void;
+  isConstantExpression(): boolean;
+  isInStrictMode(): boolean;
+
+  insertBefore(nodes: t.Node | t.Node[]): any;
+  insertAfter(nodes: t.Node | t.Node[]): any;
+  updateSiblingKeys(fromIndex: number, incrementBy: number): void;
+  unshiftContainer(listKey: string, nodes: t.Node | t.Node[]): any;
+  pushContainer(listKey: string, nodes: t.Node | t.Node[]): any;
+  hoist(scope?: Scope): void;
+
+  remove(): void;
+
+  replaceWithMultiple(nodes: t.Node[]): void;
+  replaceWithSourceString(replacement: any): void;
+  replaceWith(replacement: t.Node | t.NodePath): void;
+  replaceExpressionWithStatements(nodes: t.Node[]): t.Node;
+  replaceInline(nodes: t.Node | t.Node[]): void;
+
+  getTypeAnnotation(): t.TypeAnnotation;
+  isBaseType(baseName: string, soft?: boolean): boolean;
+  couldBeBaseType(name: string): boolean;
+  baseTypeStrictlyMatches(right: NodePath): boolean;
+  isGenericType(genericName: string): boolean;
+${isEntries.join("\n")}
+${assertEntries.join("\n")}
```

Comment:
These generated typings are more complete than the flow type annotations for NodePath. When I did a spike to get the flow type annotations into a useful state, I ran across a bunch of `NodeType.prototype` 
 trickery that flow wasn't clever enough to unpick. I ended up hand-unrolling them so that I could get on with my prototype, but this is not an approach that you would want to merge ( https://github.com/babel/babel/pull/9262/files#r244605487 ).

It may be that the simplest approach would be to generate both the flow and typescript annotations for NodePath from the same source of truth.